Framework/CurveFitting/src/Functions/CrystalFieldControl.cpp

```diff
@@ -0,0 +1,354 @@
+#include "MantidCurveFitting/Functions/CrystalFieldControl.h"
+#include "MantidCurveFitting/Functions/CrystalFieldPeaks.h"
+#include "MantidKernel/Exception.h"
+#include "MantidKernel/StringTokenizer.h"
+#include "MantidKernel/Strings.h"
+
+namespace Mantid {
+namespace CurveFitting {
+namespace Functions {
+
+using namespace API;
+
+CrystalFieldControl::CrystalFieldControl() : CompositeFunction() {
+  const bool quotedString = true;
+  // A comma-separated list of ion names
+  declareAttribute("Ions", Attribute("", quotedString));
+  // A comma-separated list of symmetry names
+  declareAttribute("Symmetries", Attribute("", quotedString));
+  // Temperature values for each spectrum.
+  declareAttribute("Temperatures", Attribute(std::vector<double>()));
+  // Default widths for peaks in each spectrum. If given it must have
+  // the same size as Temperatures or size == 1 in which case it's used
+  // for all spectra.
+  declareAttribute("FWHMs", Attribute(std::vector<double>()));
+  // Variation in FWHM of peaks when with model is used (FWHMX and FWHMY)
+  declareAttribute("FWHMVariation", Attribute(0.1));
+  // Definition of the background function
+  declareAttribute("Background", Attribute("", quotedString));
+  // Name of a IPeakFunction to use for peaks
+  declareAttribute("PeakShape", Attribute("Lorentzian"));
+  // Energy tolerance in crystal field calculations
+  declareAttribute("ToleranceEnergy", Attribute(1.0e-10));
+  // Intensity tolerance in crystal field calculations
+  declareAttribute("ToleranceIntensity", Attribute(1.0e-1));
+  // A comma-separated list of physical properties
+  declareAttribute("PhysicalProperties", Attribute(""));
+  declareAttribute("NPeaks", Attribute(0));
+  declareAttribute("FixAllPeaks", Attribute(false));
+}
+
+/// Set a value to attribute attName
+void CrystalFieldControl::setAttribute(const std::string &name,
+                                       const API::IFunction::Attribute &attr) {
+  if (name == "Ions") {
+    parseStringListAttribute("Ions", attr.asUnquotedString(), m_ions);
+  } else if (name == "Symmetries") {
+    parseStringListAttribute("Symmetries", attr.asUnquotedString(),
+                             m_symmetries);
+  } else if (name == "PhysicalProperties") {
+    parseStringListAttribute("PhysicalProperties", attr.asString(),
+                             m_physProps);
+    buildPhysPropControls();
+  } else {
+    if (name == "Temperatures") {
+      m_temperatures = attr.asVector();
+      buildControls();
+    } else if (name == "FWHMs") {
+      m_FWHMs = attr.asVector();
+      if (m_FWHMs.size() == 1 && m_FWHMs.size() != m_temperatures.size()) {
+        m_FWHMs.assign(m_temperatures.size(), m_FWHMs.front());
+      }
+    }
+    API::IFunction::setAttribute(name, attr);
```

Comment:
could we add an esle statement at the end that produces an error/warning?

Reply:
I don't think it is needed here. Temperature and FWHMs require additional actions before `IFunction::setAttribute` is called.

Reply:
I just thought it would be good to make sure that the name is valid. So is that checked elsewhere?

Reply:
Yes, by `IFunction::setAttribute`.